### PR TITLE
TW-423: Send image and file for web

### PR DESCRIPTION
**Pull request by nqhhdev**
- Issue: #423 

- Resolved: 

https://github.com/linagora/twake-on-matrix/assets/99852347/c0469a9d-344a-4fa9-b0bd-576d29e57c31

### CODE REVIEW

File: lib/presentation/extensions/send_file_web_extension.dart [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+      await client.database?.transaction(() async {
+        await client.handleSync(fakeImageEvent, direction: direction);
+      });
+    } else {
+      await client.handleSync(fakeImageEvent, direction: direction);
+    }
+  }
+
+  void clearOlderImagesCacheInRoom({String? txId}) {
+    final imageCacheQueue = getIt.get<Queue>();
```
[the remark this answers is not in the document]
Comment (nqhhdev) [reply lib/presentation/extensions/send_file_web_extension.dart:181]: Screenshot 2023-08-11 at 10 03 57

[the remark this answers is not in the document]
Comment (hoangdat) [reply lib/presentation/extensions/send_file_web_extension.dart:181]: please make sure it is singleton or factory. Be careful!

[the remark this answers is not in the document]
Comment (hoangdat) [reply lib/presentation/extensions/send_file_web_extension.dart:181]: anywhere use it also?

File: lib/domain/usecase/send_file_on_web_interactor.dart [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      final txIdMapToImageInfo =
+          await room.sendPlaceholdersWebForImages(entities: fileInfos);
+
+      for (final txId in txIdMapToImageInfo.value1.keys) {
+        await room.sendFileOnWebEvent(
+          txIdMapToImageInfo.value1[txId]!,
+          fakeImageEvent: txIdMapToImageInfo.value2[txId],
+          txid: txId,
+        );
+        room.clearOlderImagesCacheInRoom(txId: txId);
```
Comment (hoangdat) [inline lib/domain/usecase/send_file_on_web_interactor.dart:34]: what is it?

Comment (nqhhdev) [reply lib/domain/usecase/send_file_on_web_interactor.dart:34]: - Ex logic from Mr Khai.
- When we show placeHolder => remove it when is full

File: lib/domain/usecase/send_file_on_web_interactor.dart [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    required Room room,
+    required FilePickerResult filePickerResult,
+    String? txId,
+    Event? inReplyTo,
+    String? editEventId,
+    int? shrinkImageMaxDimension,
+    Map<String, dynamic>? extraContent,
+  }) async {
+    try {
+      final fileInfos = filePickerResult.files
```
Comment (hoangdat) [inline lib/domain/usecase/send_file_on_web_interactor.dart:16]:
```suggestion
      final matrixFiles = filePickerResult.files
```

File: lib/pages/chat/chat.dart [hunk 4]
```diff
@@ -344,7 +344,7 @@ class ChatController extends State<Chat> with ImagePickerMixin, SendFilesMixin {
     prefs.remove('draft_$roomId');
     var parseCommands = true;
 
-    final commandMatch = RegExp(r'^\/(\w+)').firstMatch(sendController.text);
+    final commandMatch = RegExp(r'^/(\w+)').firstMatch(sendController.text);
```
Comment (hoangdat) [inline lib/pages/chat/chat.dart:347]: why need to touch on it?

File: lib/pages/chat/events/message_content.dart [hunk 5]
```diff
@@ -130,14 +130,21 @@ class MessageContent extends StatelessWidget {
...
-              );
+            } else if (event.status == EventStatus.sending) {
+              final filePath = event.getFilePath();
+              if (filePath != null) {
+                return SendingImageWidget(
+                  filePath: filePath,
+                  event: event,
+                  onTapPreview: onTapPreview,
+                );
+              } else if (PlatformInfos.isWeb) {
```
Comment (hoangdat) [inline lib/pages/chat/events/message_content.dart:141]: why not separate the widget? `SendImageInfoWidget` and `SendImageBlobWidget`

File: lib/presentation/extensions/send_file_web_extension.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (hoangdat) [reply lib/presentation/extensions/send_file_web_extension.dart:181]: the good way: register it with a name. And I think it should be a `factory`.

File: lib/presentation/extensions/send_file_web_extension.dart [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+      if (sendingFilePlaceholders.containsKey(txId)) {
+        sendingFilePlaceholders.remove(txId);
+      }
+      if (sendingFileThumbnails.containsKey(txId)) {
+        sendingFileThumbnails.remove(txId);
+      }
+    }
+  }
+
+  Future<
```
[the remark this answers is not in the document]
Comment (hoangdat) [reply lib/presentation/extensions/send_file_web_extension.dart:193]: ??

File: lib/presentation/extensions/send_file_web_extension.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (hoangdat) [reply lib/presentation/extensions/send_file_web_extension.dart:181]: @sherlockvn WDYT?

[the remark this answers is not in the document]
Comment (sherlockvn) [reply lib/presentation/extensions/send_file_web_extension.dart:181]: yes factory is oke

File: lib/presentation/extensions/send_file_web_extension.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (sherlockvn) [reply lib/presentation/extensions/send_file_web_extension.dart:193]: or use class like @hoangdat suggest

File: lib/pages/chat/events/send_image_blob_widget.dart [hunk 7]
```diff
@@ -0,0 +1,86 @@
+// ignore_for_file: unused_import
```
Comment (drminh2807) [inline lib/pages/chat/events/send_image_blob_widget.dart:1]: why need it ?

File: lib/pages/chat/events/send_image_info_widget.dart [hunk 8]
```diff
@@ -1,20 +1,23 @@
+// ignore_for_file: unused_import
```
Comment (drminh2807) [inline lib/pages/chat/events/send_image_info_widget.dart:1]: same

**Review by sherlockvn — APPROVED** [review]

**Review by sherlockvn — APPROVED** [review]

File: lib/utils/matrix_sdk_extensions/event_extension.dart [hunk 9]
```diff
@@ -73,6 +73,11 @@ extension LocalizedBody on Event {
     return room.sendingFilePlaceholders[txId];
   }
 
+  Uint8List? getBytes() {
+    final txId = unsigned?['transaction_id'] ?? eventId;
+    return room.sendingFilePlaceholders[txId]?.bytes;
+  }
+
```
Comment (hoangdat) [inline lib/utils/matrix_sdk_extensions/event_extension.dart:80]: we need it?

Comment (nqhhdev) [reply lib/utils/matrix_sdk_extensions/event_extension.dart:80]: I'll remote it, it not use

**Review by hoangdat — APPROVED** [review]

**Pull request merged**